### ISSUE TITLE
Race conditin in QEventDispatcher on OS/2

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/qtbase-os2#51]
It appears that there is a race condition on an auxiliary thread within `QEventDispatcherOS2` internals that calls `select()`: this call is supposed to be interrupted (e.g. to pick up new sockets) with a `so_cancel()` call. However, if this call is issued moments before `select` actually enters its wait state, nothing happens. Consider this sequence of events:
```
main_thread: start select_thread
...
main_thread: receive register new socket request
select_thread: lock mutex
main_thread: wait for mutex (locked by select_thread)
select_thread: pick up new sockets
select_thread: unlock mutex [0]
main_thread: lock mutex [1]
main_thread: call so_cancel to abort select [2]
select_thread: call select [3]
```
Execution may happen in any of these orders: [123], [132], [312] after the mutex is unlocked at [0]. The only guarantee here is that [2] obviously comes ater [1] (since both calls are on the same thread). However, the only working solution for us is when [2] comes after [3] (i.e. when `so_cancel` gets called after `select`). This equals to orders [132] and [312]. However, due to true multithreading (and especially SMP multithreading), order [123] is also very likely. And when it happens, we are screwed. I've found it when working on `tst_QSocketNotifier` and `tst_QTimer` testcases within #29. Can reproduce it each 2 of 5 runs or so.

Note that the very same problem exists in Qt4. However the code is a bit different there so a kick in of [1] and [2] before [0] and [3] is just less likely. And frankly, there is no universal solution here at first sight which is bad. We need to find one.

**Comment by dmik** [comment]
I wrote a simple test case and this test case works exactly as advertised: even if I explicitly `so_cancel` before starting a thread that does `select` on that socket handle, `select` immediately fails with EINTR. So somehow the Qt build affects the way TCP/IP stack works. This totally confuses me.

**Comment by dmik** [comment]
I also wrote a simple Qt test case where I simply create two QSocketNotifier instances for a socketpair. And it also works. Hmmm. So it must be something else involved in `tst_QSocketNotifier` that breaks select. May be it's manually created local sockets. I have to reproduce the same thing in my Qt test case.

**Comment by dmik** [comment]
Okay, it turned out that the problem was a regression brought by the last version of my QSocketNotifier implementation (which I did within #50). Requests to select new sockets from the main thread were occasionally overwritten by the results of the previous operation on the select thread. This led to the wrong socket handle to be used in `so_cancel`. This, of course, did not interrupt `select` because that socket was missing from its current working set.

Anyway, I fixed this now and tst_QSocketNotifier from #29 is rock solid. I found that we have problems with zero timers, but since it has nothing to do with sockets, I created #52 for it. Closing this one.
